### PR TITLE
Makes psychic link usable while moving

### PR DESCRIPTION
**Pull request by Walarks**
## About The Pull Request
Makes psychic link channel-able while moving.

## Why It's Good For The Game
Xenos don't stand still when gorger is using the ability on them and the ability ends up being underused.

## Changelog
:cl:
balance: psychic link is now channel-able while moving
fix: fixes signal loop in psychic link damage redirect
/:cl:

### CODE REVIEW

File: code/modules/mob/living/carbon/xenomorph/castes/gorger/abilities_gorger.dm [hunk 1]
```diff
@@ -291,25 +295,38 @@
 		if(!silent)
 			to_chat(owner, span_notice("It is beyond our reach, we must be close and our way must be clear."))
 		return FALSE
-	if(!do_mob(owner, target, GORGER_PSYCHIC_LINK_CHANNEL, BUSY_ICON_FRIENDLY, BUSY_ICON_MEDICAL))
+	if(!do_mob(owner, target, GORGER_PSYCHIC_LINK_CHANNEL, BUSY_ICON_FRIENDLY, BUSY_ICON_MEDICAL, ignore_flags = IGNORE_LOC_CHANGE, extra_checks = new/datum/callback/(src, .proc/channel_checks, target)))
+		if(!silent)
+			to_chat(owner, span_warning("The link was interrupted and we need to recover before another attempt."))
+		add_cooldown(10 SECONDS)
```
Comment (BraveMole) [inline code/modules/mob/living/carbon/xenomorph/castes/gorger/abilities_gorger.dm:301]: ooldown on fail is usually shit

File: code/modules/mob/living/carbon/xenomorph/damage_procs.dm [hunk 2]
```diff
@@ -169,27 +169,31 @@
...
 /mob/living/carbon/xenomorph/adjustFireLoss(amount, updating_health = FALSE)
-
 	var/list/amount_mod = list()
 	SEND_SIGNAL(src, COMSIG_XENOMORPH_BURN_DAMAGE, amount, amount_mod)
 	for(var/i in amount_mod)
 		amount -= i
+	adjustFireLossDirect(amount, updating_health)
 
+///Adjusts fireloss without sending a signal. Useful when it can cause a signal loop
+/mob/living/carbon/xenomorph/proc/adjustFireLossDirect(amount, updating_health = FALSE)
```
Comment (BraveMole) [inline code/modules/mob/living/carbon/xenomorph/damage_procs.dm:196]: uuuh, what is this??

**Review by BraveMole — CHANGES_REQUESTED** [review]

File: code/modules/mob/living/carbon/xenomorph/damage_procs.dm (hunk 2 above)
Comment (Walarks) [reply code/modules/mob/living/carbon/xenomorph/damage_procs.dm:196]: It modifies fireloss without sending the signal, so the action doesn't trigger effects that activate on being damaged and such. Should I make it a define?

Comment (BraveMole) [reply code/modules/mob/living/carbon/xenomorph/damage_procs.dm:196]: that's extremly iffy. If you send a signal when an action is done, you do it all the time. You don't have one proc that does a signal and the other not, this is awful design

**Review by BraveMole — CHANGES_REQUESTED** [review]

File: code/modules/mob/living/carbon/xenomorph/castes/gorger/abilities_gorger.dm [hunk 3]
```diff
@@ -291,26 +298,49 @@
...
+	if(HAS_TRAIT(target, TRAIT_PSY_LINKED))
+		to_chat(owner, span_notice("[target] is already linked to a xenomorph."))
 		return FALSE
 	return TRUE
 
+///Runs checks while the status is being applied
+/datum/action/xeno_action/activable/psychic_link/proc/channel_checks(atom/target)
+	if(get_dist(owner, target) > GORGER_PSYCHIC_LINK_RANGE)
+		return FALSE
+	return TRUE
```
Comment (BraveMole) [inline code/modules/mob/living/carbon/xenomorph/castes/gorger/abilities_gorger.dm:314]:
```suggestion
	return (get_dist(owner, target) <= GORGER_PSYCHIC_LINK_RANGE)
```

File: code/datums/status_effects/xeno_buffs.dm [hunk 4]
```diff
@@ -101,34 +100,42 @@
...
+	. = ..()
 	owner = new_owner
 	duration = set_duration
 	src.target_mob = target_mob
 	src.link_range = link_range
 	src.redirect_mod = redirect_mod
 	src.minimum_health = minimum_health
-	RegisterSignal(owner, COMSIG_MOVABLE_MOVED, .proc/handle_dist)
-	RegisterSignal(target_mob, COMSIG_MOVABLE_MOVED, .proc/handle_dist)
+	instance_id = "[src][owner][target_mob]"
```
Comment (BraveMole) [inline code/datums/status_effects/xeno_buffs.dm:114]: what's the point of that?

**Review by BraveMole — CHANGES_REQUESTED** [review]

File: code/datums/status_effects/xeno_buffs.dm [hunk 5]
```diff
@@ -102,33 +101,43 @@
...
+/datum/status_effect/xeno_psychic_link/on_creation(mob/living/new_owner, set_duration, mob/living/carbon/target_mob, link_range, redirect_mod, minimum_health, scaling = FALSE)
 	owner = new_owner
 	duration = set_duration
 	src.target_mob = target_mob
 	src.link_range = link_range
 	src.redirect_mod = redirect_mod
 	src.minimum_health = minimum_health
-	RegisterSignal(owner, COMSIG_MOVABLE_MOVED, .proc/handle_dist)
-	RegisterSignal(target_mob, COMSIG_MOVABLE_MOVED, .proc/handle_dist)
+	ADD_TRAIT(target_mob, TRAIT_PSY_LINKED, REF(src))
```
Comment (BraveMole) [inline code/datums/status_effects/xeno_buffs.dm:116]: why a ref and not a defined string like every other trait

File: code/datums/status_effects/xeno_buffs.dm [hunk 6]
```diff
@@ -102,33 +101,43 @@
...
-	target_mob.add_filter(TARGET_ID, 2, outline_filter(2, PSYCHIC_LINK_COLOR))
-	var/link_message = "[owner] has linked to you and is redirecting some of your injuries. If they get too hurt, the link may be broken. "
+	var/link_message = "[owner] has linked to you and is redirecting some of your injuries. If they get too hurt, the link may be broken."
 	if(link_range > 0)
-		link_message += "Keep within [link_range] tiles to maintain it."
+		link_message += " Keep within [link_range] tiles to maintain it."
+		RegisterSignal(owner, COMSIG_MOVABLE_MOVED, .proc/handle_dist)
+		RegisterSignal(target_mob, COMSIG_MOVABLE_MOVED, .proc/handle_dist)
 	to_chat(target_mob, span_xenonotice(link_message))
+	owner.add_filter(REF(src), 2, outline_filter(2, PSYCHIC_LINK_COLOR))
```
Comment (BraveMole) [inline code/datums/status_effects/xeno_buffs.dm:128]: why a ref

**Review by BraveMole — CHANGES_REQUESTED** [review]

File: code/datums/status_effects/xeno_buffs.dm (hunk 5 above)
Comment (Walarks) [reply code/datums/status_effects/xeno_buffs.dm]: I was thinking of making it possible to apply psychic link if you don't have the buff on yourself, but are the target of one, meaning you could have a trait from your buff and from someone else's, but I removed this option now.

File: code/datums/status_effects/xeno_buffs.dm (hunk 6 above)
Comment (Walarks) [reply code/datums/status_effects/xeno_buffs.dm]: changed

**Review by BraveMole — APPROVED** [review]

**Pull request merged**